### PR TITLE
Add ZStream#throttleShape, ZStream#throttleShapeM

### PR DESCRIPTION
**Pull request by vasilmkd**
Part of #1027

### CODE REVIEW

**Comment by vasilmkd** [comment]
Can anything be done about the failing tests? I believe they are unrelated.

**Comment by iravid** [comment]
Hey Vasil, this looks great. I’ll have time to review this tomorrow. I’d just like to note that you don’t have to use RefM; you can use Ref and run the delay after modify has finished.

File: streams/jvm/src/test/scala/zio/stream/SinkSpec.scala [hunk 1]
```diff
@@ -225,4 +228,55 @@ class SinkSpec(implicit ee: org.specs2.concurrent.ExecutionEnv)
       } yield test
     }
   }
+
+  private def throttleShape = {
+
+    def sinkTest(sink: ZSink[Clock, Nothing, Nothing, Int, Int]) =
```
Comment (iravid) [inline streams/jvm/src/test/scala/zio/stream/SinkSpec.scala]: very nice use of the TestClock!

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 2]
```diff
@@ -1128,4 +1134,59 @@ object ZSink extends ZSinkPlatformSpecific {
...
+                        else {
+                          val c = missing / units
+                          if (c * units < missing) c + 1 else c
+                        }
+                      val newTokens    = tokens + cycles * units - weight
+                      val newTimestamp = timestamp + cycles * duration.toNanos
+                      val delay        = Duration.Finite(newTimestamp - current)
+                      (delay, (newTokens, newTimestamp))
+                  }
+          _ <- state._2.succeed(a).delay(delay)
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala]: `delay` forces a bounce through the platform's scheduler; can we skip it if `delay <= 0`? You can also do `clock.sleep(delay)` directly, no need to delay a value.

Comment (vasilmkd) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: Done.

File: streams/jvm/src/test/scala/zio/stream/SinkSpec.scala (hunk 1 above)
Comment (vasilmkd) [reply streams/jvm/src/test/scala/zio/stream/SinkSpec.scala]: Thanks. And thank you for the help with the sbt setup.

**Comment by iravid** [comment]
Amazing work! Thank you @vasilmkd!

**Pull request merged**